lib/APPEndlessRun/DrivingState.cpp

```diff
@@ -0,0 +1,455 @@
+    const uint16_t LINE_SENSOR_OFF_TRACK_VALUE = 200;
+
+    /* Note, the start-/end line detection must be debounced. Otherwise
+     * especially in low speed use cases, the line may be in one cycle
+     * detected, in the next not and then detected again. This would lead
+     * to a start line detection and afterwards to a end line detection,
+     * which would be wrong.
+     *
+     * Note the three sensors in the middle are handled as one sensor to
+     * avoid detection problems with different kind of line widths.
+     */
+    if ((LINE_SENSOR_OFF_TRACK_VALUE <= leftSensor) && (LINE_SENSOR_OFF_TRACK_VALUE <= middleSensor) &&
+        (LINE_SENSOR_OFF_TRACK_VALUE <= rightSensor))
+    {
+        if (DEBOUNCE_CNT > m_startEndLineDebounce)
+        {
+            ++m_startEndLineDebounce;
+        }
+
+        if (DEBOUNCE_CNT <= m_startEndLineDebounce)
+        {
+            /* Reset orientation, clear position and PID controller errors. */
+            Odometry::getInstance().setOrientation(FP_PI() / 2);
+            Odometry::getInstance().clearPosition();
+            m_pidCtrl.clear();
+        }
+        else
+        {
+            isDetected = false;
+        }
+    }
+    else
+    {
+        m_startEndLineDebounce = 0;
+    }
+
+    return isDetected;
+}
+
+bool DrivingState::isTrackGapDetected(int16_t position) const
+{
+    bool                isDetected  = false;
+    const ILineSensors& lineSensors = Board::getInstance().getLineSensors();
+
+    /* Position value after loosing the track and sensor 0 saw it as last.
+     * It depends on the Zumo32U4LineSensors::readLine() implementation.
+     */
+    const int16_t POS_MIN = 0;
+
+    /* Position value after loosing the track and sensor N saw it as last.
+     * It depends on the Zumo32U4LineSensors::readLine() implementation.
+     */
+    const int16_t POS_MAX = (lineSensors.getNumLineSensors() - 1) * 1000;
+
+    /* Note, no debouncing is done here. If necessary, it shall be done
+     * outside this method.
+     */
+    if ((POS_MIN >= position) || (POS_MAX <= position))
+    {
+        isDetected = true;
+    }
+
+    return isDetected;
+}
+
+void DrivingState::adaptDriving(int16_t position)
+{
+    DifferentialDrive&  diffDrive       = DifferentialDrive::getInstance();
+    const ILineSensors& lineSensors     = Board::getInstance().getLineSensors();
+    const int16_t       MAX_MOTOR_SPEED = diffDrive.getMaxMotorSpeed();
+    int16_t             speedDifference = 0; /* [steps/s] */
+    int16_t             leftSpeed       = 0; /* [steps/s] */
+    int16_t             rightSpeed      = 0; /* [steps/s] */
+
+    /* Our "error" is how far we are away from the center of the
+     * line, which corresponds to position (max. line sensor value multiplied
+     * with sensor index).
+     *
+     * Get motor speed difference using PID terms.
+     */
+    speedDifference = m_pidCtrl.calculate(lineSensors.getSensorValueMax() * 2, position);
```

Comment:
If you stop the robot and the robot is not exactly on line, the integral and derivative part will increase till limit. Thats bad.